File: 279. Perfect Squares.cpp
class Solution {
public:
    int numSquares(int n) {
        vector <int> T (n + 1, n + 1);
        T[0] = 0;
        T[1] = 1;
        for (int i = 1; i <= n; i++)
        {
            for (int j = 1; j * j <= i; j++)
                T[i] = T[i] < T[i - j * j] + 1 ? T[i] : T[i - j * j] + 1;
        }
        return T[n];
    }
};

File: 279. Perfect Squares.py
# timeout

class Solution(object):
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        T = [n+1 for i in xrange(n+1)]
        for i in xrange(1, n + 1):
            if i*i <= n:
                T[i*i] = 1
        for i in xrange(1, n + 1):
            for j in xrange(1, i):
                T[i] = min(T[i], T[i-j] + T[j])
        return T[n]
